### PR TITLE
[TVMC] --disable-pass option added to compile mode

### PR DESCRIPTION
**Pull request by d-smirnov**
Added --disable-pass option to TVMC compile mode to disallow certain supplied passes in PassContext for the compiler.

### CODE REVIEW

File: python/tvm/driver/tvmc/common.py [hunk 1]
```diff
@@ -333,6 +333,29 @@ def tracker_host_port_from_cli(rpc_tracker_str):
     return rpc_hostname, rpc_port
 
 
+def parse_disabled_pass(input_string):
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:336]: This utility could be more general. Since its functionatlity is taking a list of passes in a string and checking if they are valid, we don't need to limit this function to "disabled" passes. Instead, we could name it `parse_pass_list_str`, which does the same thing as now.

File: python/tvm/driver/tvmc/common.py [hunk 2]
```diff
@@ -333,6 +333,29 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+    Parameters
+    ----------
+    input_string: str
+        Possibly comma-separated string with the names of disabled passes
+
+    Returns
+    -------
+    list: a list of disabled passes.
+    """
+    if input_string is not None:
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:348]: It's weird to see an only input could be None. I would suggest input must be a string, and you could assign `""` to the default value in the argparser.

File: python/tvm/driver/tvmc/common.py [hunk 3]
```diff
@@ -333,6 +333,29 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+        pass_list = input_string.split(",")
+        nf = [_ for _ in pass_list if tvm.get_global_func("relay._transform." + _, True) is None]
+        if len(nf) > 0:
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:351]:
```suggestion
        invalid_passes = [p for p in pass_list if tvm.get_global_func("relay._transform.%s" % p, True) is None]
        if len(invalid_passes) > 0:
```
Don't use underline as the variable name if you actually refer it.

File: python/tvm/driver/tvmc/compiler.py [hunk 4]
```diff
@@ -95,6 +95,12 @@ def add_compile_parser(subparsers):
         type=common.parse_shape_string,
         default=None,
     )
+    parser.add_argument(
+        "--disabled-pass",
```
Comment (comaniac) [inline python/tvm/driver/tvmc/compiler.py:99]: I would prefer the name `--disable-pass` in the CLI (active verb + noun).

File: python/tvm/driver/tvmc/compiler.py [hunk 5]
```diff
@@ -95,6 +95,12 @@ def add_compile_parser(subparsers):
...
+        help="disable specific passes, comma-separated list of pass names",
+        type=common.parse_disabled_pass,
+        default=None,
```
Comment (comaniac) [inline python/tvm/driver/tvmc/compiler.py:102]: As I mentioned in another comment, you could assign `""` to default, so that the utility function can take only string but not None.

File: python/tvm/driver/tvmc/compiler.py [hunk 6]
```diff
@@ -138,6 +146,8 @@ def compile_model(
     target_host=None,
     tuning_records=None,
     alter_layout=None,
+    shape_dict=None,
+    disabled_pass=None,
```
Comment (comaniac) [inline python/tvm/driver/tvmc/compiler.py:150]: Update the docstring. And `shape_dict` seems not related to this PR?

**Review by comaniac — CHANGES_REQUESTED** [review]
Please also add test cases.

**Comment by comaniac** [comment]
Also cc @leandron

File: python/tvm/driver/tvmc/compiler.py [hunk 7]
```diff
@@ -95,6 +95,12 @@ def add_compile_parser(subparsers):
         type=common.parse_shape_string,
         default=None,
     )
+    parser.add_argument(
+        "--disabled-pass",
+        help="disable specific passes, comma-separated list of pass names",
```
Comment (leandron) [inline python/tvm/driver/tvmc/compiler.py:100]: Is there a way to obtain a list of all valid passes so that we can show in the help option, or is it too many?

File: python/tvm/driver/tvmc/common.py [hunk 8]
```diff
@@ -333,6 +333,29 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+            raise argparse.ArgumentTypeError(
+                "Following passes are not registered within tvm: " + str(nf)
+            )
+        return pass_list
+    return None
```
Comment (leandron) [inline python/tvm/driver/tvmc/common.py:356]: Once `if input_string is not None:` is removed, this can be removed as well.

**Review by leandron — COMMENTED** [review]
Please add a some test cases.

**Review by comaniac — APPROVED** [review]
LGTM. Please evaluate @leandron's comment of listing all available passes. We could skip that if it's too many as suggested.

File: python/tvm/driver/tvmc/compiler.py (hunk 7 above)
Comment (d-smirnov) [reply python/tvm/driver/tvmc/compiler.py:100]: Added displaying of available passes in case if supplied pass is unknown to system

**Review by leandron — APPROVED** [review]
Considering it is an advanced feature, and it is quite a big list of passes, I agree with your choice of showing it only when the user points to some pass that doesn't exist. LGTM.

**Comment by comaniac** [comment]
Thanks @d-smirnov @leandron
